lib/puppet/util/execution.rb

```diff
@@ -150,6 +160,9 @@ def self.execute(command, options = NoOptionsSpecified)
         :gid => nil,
         :combine => NoOptionsSpecified.equal?(options),
         :stdinfile => nil,
+        :stdin_file => nil,
+        :stdin_string => nil,
+        :stdin_yield => false,
```

Comment:
@PaulSD I'd prefer that the execute method just takes an `IO` for stdin, which could either be a `StringIO` or `File`.

The yield logic makes the implementation much more complicated. To support that, we should add a different `execute` method that yields `stdin`, `stdout`, `stderr` like popen. But we are planning on reworking the execute API in the very near future, so let's not add yield support right now.

Reply:
@joshcooper Actually ... accepting `StringIO` doesn't work the way you probably expect.  Unlike all other `IO` objects, `StringIO` does not wrap an OS-level file handle, so you can't simply pass a `StringIO` object into the process as `stdin` (See https://www.ruby-forum.com/t/stringio-and-reopen/178396 for posix, and https://github.com/chef/win32-process/blob/ffi/lib/win32/process.rb#L441 for windows).

It is still possible to replace the separate stdin_whatever arguments with a single argument that can either be a StringIO or a File, simply so that the object type can be used to determine how the argument should be handled.  However, if it is a StringIO, I will still need to create an IO.pipe, then stream the StringIO contents into the IO.pipe (maybe using `IO.copy_stream`).